Add overlay modes for metadata; overwrite, add missing etc.

mizaki opened the pull request:
This is the barebones of some overlay options. It's missing the settings which I will add if you are happy with the general idea. It's also missing any tests for the new options. I have done nothing with the `notes` as yet.

When it comes to the notes, using `CVDB xxxxx` will still work with Mylar3 and as long as `Issue ID` is not used (such as `Metron Issue ID` and instead `Metron ID`) then the multi-source `Tagged with CT etc. etc. CVDB xxxxx, GCD xxxxxx, Metron xxxxx` will not cause problems. I've checked Komga, Kavita and LANraragi and none appear to use the `notes` field.

mizaki committed:
Add OverlayMode options for read style and data source

mizaki committed:
Add OverlayMode options for read style and data source

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -103,6 +103,13 @@ class GenericMetadata:
     editor_synonyms = ("editor", "edits", "editing")
     _translator_synonyms = ("translator", "translation")
 
+    overlay_original = 0
+    overlay_overwrite = 1
+    overlay_overwrite_no_clear = 2
+    overlay_add_missing = 3
+    overlay_combine_new = 4
+    overlay_combine_cur = 5
```
Make an enum class for this. At a minimum.

mizaki replied:
There is the Enum https://docs.python.org/3/library/enum.html which I looked at but thought I'd replicate what had been done in `IssueIdentifier` etc. Should I use the enum class or just make a normal class?

lordwelch replied:
Use the enum class. ComicTagger was started back on python 2.7 and still has some leftover concepts from back then

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,25 +209,112 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
-
-        That is, when the new object has non-None values, over-write them
-        to this one.
+    def casefold_check(self, a: str, b: str) -> bool:
```
This doesn't help anything. remove it and use the comparison directly.

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,25 +209,112 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+    def casefold_check(self, a: str, b: str) -> bool:
+        return a.casefold() == b.casefold()
+
+    def overlay(self, new_md: GenericMetadata, mode: int = 0) -> None:
```
Never use "magic numbers" in any functions always use a named variable. It shouldn't matter if this is a int, float, str or a class this should be a name to refer to the mode that happens.

lordwelch replied:
Here's an example of me doing it with strings and breaking something
https://github.com/comictagger/comictagger/blob/09490b8ebf70415211dc856663ea1c1fd258b47b/comictaggerlib/ctsettings/plugin.py#L65
It should be `Archive` is it is set here
https://github.com/comictagger/comictagger/blob/09490b8ebf70415211dc856663ea1c1fd258b47b/comictaggerlib/ctsettings/plugin.py#L100

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,25 +209,112 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
         def assign(cur: str, new: Any) -> None:
-            if new is not None:
-                if isinstance(new, str) and len(new) == 0:
+            if mode == self.overlay_original:
```
Don't do this. You are going 8-levels deep on some of these. At a minimum make each of these separate functions that get called here.

You could also just make the mode argument the actual function so you don't even need the if chain

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -278,20 +372,65 @@ def assign(cur: str, new: Any) -> None:
 
-    def overlay_credits(self, new_credits: list[Credit]) -> None:
-        if isinstance(new_credits, str) and len(new_credits) == 0:
+    def overlay_credits(self, new_credits: list[Credit], mode: int = 0) -> None:
```
Same as above with assign

mizaki commented:
I've tried to simplify things with https://github.com/comictagger/comictagger/pull/598/commits/ddbc5d83f8d65c0dbb5687b92ca400ef4202471f (just the `assign_` changes for now) I didn't know if it's better check if a string is 0 rather that just set the `cur` to whatever `new` is even if it was an empty string?

mizaki commented:
When it comes to saving the setting, is it best to use the Enum number or name?

lordwelch commented:
> When it comes to saving the setting, is it best to use the Enum number or name?

parser.add_setting("value", default=ENUM.VALUE, type=ENUM, cmdline=False)

This won't work with int values with our current file loading code.
We would need ` or isinstance(default, Enum) or (isinstance(setting.type, type) and issubclass(setting.type, Enum))` added to https://github.com/comictagger/comictagger/blob/09490b8ebf70415211dc856663ea1c1fd258b47b/comictaggerlib/ctsettings/__init__.py#L58

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,24 +212,122 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+            new.difference_update(remove_set)
+            getattr(self, cur).update(new)
+
+    def assign_original(self, cur: str, new: Any) -> None:
```
This should be `assign_overlay`

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,24 +212,122 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+            else:
+                setattr(self, cur, new)
+
+    def assign_overwrite(self, cur: str, new: Any) -> None:
```
If everything is being overwritten isn't it just making a copy of the new metadata object?

mizaki replied:
I don't know if it was supposed to do something else originally but yes, it will now overwrite everything. I think this should then replace the `Clear all existing metadata before import...` so everything is in one location/setting?

lordwelch replied:
We can look at that after this pr

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,24 +212,122 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+        else:
+            setattr(self, cur, new)
+
+    def assign_overwrite_no_clear(self, cur: str, new: Any) -> None:
```
What's the point in this? This also allows a string to be assigned to a list

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,24 +212,122 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+            setattr(self, cur, new)
+
+    def assign_combine_new(self, cur: str, new: Any) -> None:
+        # Combine lists and sets and prefer 'new' for strings, floats and ints
```
So overlay + merge list/set?

mizaki replied:
Yes (mostly?). It will merge lists/sets and replace all other values with `new` but won't clear a list/set on an empty string. Is "merge" the preferred nomenclature?

lordwelch replied:
Just wanted to make sure I new what it was trying to do

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,24 +212,122 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+                self.assign_dedupe_new(new, cur)
+            elif isinstance(new, str) and len(new) > 0:
+                setattr(self, cur, new)
+            elif isinstance(new, (int, float)) and new > 0:
+                setattr(self, cur, new)
```
Why is 0 ignored? `None` is used for empty. This makes it impossible to update an issue number 0 such as https://www.dc.com/comics/batman-2011/batman-0

mizaki replied:
That makes issue number a special case? Is there anything else that could be considered that you'd want to overwrite a value `> 0` with `0`? 

`None` is for empty but a talker might return a 0 for `null/None`, is the talker expected to decide if `0` should be `None`?

lordwelch replied:
Yes, if the value is different from the default (None for most types) it's assumed that it was set on purpose

lordwelch replied:
The current condition also disallows setting any negative numbers. The whole reason int values default to None is so that all integer values can be used

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,24 +212,122 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+                setattr(self, cur, new)
+            elif isinstance(new, (list, set)) and len(new) > 0 and isinstance(getattr(self, cur), (list, set)):
+                self.assign_dedupe_cur(new, cur)
+            elif isinstance(new, (int, float)) and getattr(self, cur) == 0:
```
This won't update anything 90% of the time int attributes do not start-off as 0 they start as `None`

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,24 +212,122 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+        """Overlay a new metadata object on this one
+
+        Modes:
+        0. Original behaviour: That is, when the new object has non-None values, over-write them to this one.
```
Don't use numbers here

mizaki commented:
Is https://github.com/comictagger/comictagger/pull/598/commits/75c825be1d880e740f3b562a66b7b6b6c07e2cd5 in the right area with the class?

lordwelch reviewed (commented):
Add a single test for each mode that covers each type so that we know what it actually does, it will also make some of the mistakes a little more obvious.

Comments are in the wrong place now but still apply.

mizaki commented:
I'm trying to understand the current overlay:
``` python
    def assign_overlay(self, cur: str, new: Any) -> None:
        """Overlay - When the new object has non-None values, over-write them to this one."""
        if new is not None:
            if isinstance(new, str) and len(new) == 0:
                if isinstance(getattr(self, cur), (list, set)): # If an empty string is passed to a list/set field then it is cleared
                    getattr(self, cur).clear()
                else:
                    setattr(self, cur, None) # Otherwise it's any empty string in a string field and should set to None
            elif isinstance(new, (list, set)) and len(new) == 0: # If an empty list/set is passed, do nothing. Why does this not clear the field the same as an empty string?
                pass
            else:
                setattr(self, cur, new) # No type checking, just set the `cur` to `new`
```
Should an empty list/set not `clear` it? Am I missing something?

And in general, should `assign` be doing type checking?

lordwelch commented:
An empty list is the default

mizaki commented:
> An empty list is the default

But should a list/set expect an empty string? That should be a type error?

mizaki commented:
I've added some tests and given each one a doc string so hopefully it makes more sense now. Credit tests are left to do.

lordwelch commented:
> But should a list/set expect an empty string? That should be a type error?

I believe this was set in order to make the `-m` option work for clearing lists

mizaki commented:
I had to duplicate the `StrEnum` code from `resulttype.py` as using it caused an import loop. I guess the best thing to do is move it to a new/another file, if so, which one?

lordwelch commented:
> I had to duplicate the `StrEnum` code from `resulttype.py` as using it caused an import loop. I guess the best thing to do is move it to a new/another file, if so, which one?

Copying it is fine, it'll get removed eventually when CT no longer supports 3.10

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,24 +252,129 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+            elif getattr(self, cur) is None:
+                setattr(self, cur, new)
+
+    def overlay(self, new_md: GenericMetadata, mode: OverlayMode = OverlayMode.overlay) -> None:
+        """Overlay a new metadata object on this one"""
+
+        assign_funcs = {}
+        for mem in OverlayMode:
+            assign_funcs[mem.name] = getattr(self, "assign_" + mem.name)
 
         def assign(cur: str, new: Any) -> None:
```
There's a whole lot of processing going on here that doesn't need to happen. This dict can just be manually created.

Instead of doing any of that how about this
```python
def assign_overlay(md: GenericMetadata, attribute: str, new_value: Any) -> None:
    """Overlay - When the new object has non-None values, over-write them to this one."""
    if new_value is not None:
        if isinstance(new_value, str) and len(new_value) == 0:
            if isinstance(getattr(md, attribute), (list, set)):
                getattr(md, attribute).clear()
            else:
                setattr(md, attribute, None)
        elif isinstance(new_value, (list, set)) and len(new_value) == 0:
            pass
        else:
            setattr(md, attribute, new_value)
...
class GenericMetadata:
    ...
    def overlay(self, new_md: GenericMetadata, assign=assign_overlay) -> None:
        """Overlay a new metadata object on this one"""

        if not new_md.is_empty:
            self.is_empty = False

        assign(self, "tag_origin", new_md.tag_origin)
```
Or instead of keeping the current function signature
```python
    def overlay(self, new_md: GenericMetadata, assign=assign_overlay) -> None:
        """Overlay a new metadata object on this one"""

        if not new_md.is_empty:
            self.is_empty = False

        self.tag_origin = assign(self.tag_origin, new_md.tag_origin)
```

lordwelch replied:
This (`self.tag_origin = assign(self.tag_origin, new_md.tag_origin)`) now makes assign not representative of what it is doing so should be renamed if used

mizaki replied:
``` python
    def overlay(self, new_md: GenericMetadata, assign=assign_overlay) -> None:
        """Overlay a new metadata object on this one"""

        if not new_md.is_empty:
            self.is_empty = False

        self.tag_origin = assign(self.tag_origin, new_md.tag_origin)
```
The `overlay` gets called from outside and will pass in the `config[0].Comic_Book_Lover__metadata_overlay` for the `assign` so keeping `def overlay(self, new_md: GenericMetadata, mode: OverlayMode = OverlayMode.overlay) -> None:` makes the most sense? Along with:
```python
assign_funcs = {
            "overlay": self.assign_overlay,
            "overwrite": self.assign_overwrite,
            "overwrite_no_clear": self.assign_overwrite_no_clear,
            "add_missing": self.assign_add_missing,
            "combine_new": self.assign_combine_new,
            "combine_current": self.assign_combine_current,
        }

        def assign(cur: str, new: Any) -> None:
            assign_func = assign_funcs.get(mode.name, self.assign_overlay)
            assign_func(cur, new)
```

`self.tag_origin = assign(self.tag_origin, new_md.tag_origin)`
I did think having a return would be easier now there is multiple options, so I will change it to that.

mizaki replied:
> This (`self.tag_origin = assign(self.tag_origin, new_md.tag_origin)`) now makes assign not representative of what it is doing so should be renamed if used

Should `tag_origin` be changed to a `set[tag_origin]`?

lordwelch replied:
> Should `tag_origin` be changed to a `set[tag_origin]`?

Let's leave it a list for now



> ```python
> assign_funcs = {
>             "overlay": self.assign_overlay,
>             "overwrite": self.assign_overwrite,
>             "overwrite_no_clear": self.assign_overwrite_no_clear,
>             "add_missing": self.assign_add_missing,
>             "combine_new": self.assign_combine_new,
>             "combine_current": self.assign_combine_current,
>         }
> ```

You're so close now replace these strings with the enums
`"overlay" -> OverlayMode.overlay`

lordwelch replied:
And then the assign function you can change to
```python
assign = assign_funcs.get(mode, self.assign_overlay)
```

mizaki replied:
With `tag_origin`; change it to a list and also add date/time to the `TagOrigin` class? Then it gives options for merging etc.

lordwelch replied:
Adding a date doesn't make sense as there is nothing that would actually set it in a useable way. CT would set it and then it would save it and then it would disappear.

lordwelch commented on tests/genericmetadata_test.py:
```diff
@@ -22,6 +23,78 @@ def test_metadata_overlay(md: comicapi.genericmetadata.GenericMetadata, replaced
     assert md == expected
 
 
+def test_metadata_overlay_overwrite():
+    md = comicapi.genericmetadata.GenericMetadata(series="test", issue="1", title="test", genres={"test", "test2"})
+    overwrite = comicapi.genericmetadata.GenericMetadata(
+        series="test2", issue="2", title="test2", genres={"test3", "test4"}
+    )
+    md.overlay(overwrite, OverlayMode.overwrite)
+
+    assert md == overwrite
```
This test makes it obvious that this mode is useless. Just use md.copy()

lordwelch replied:
As in anywhere that would have called `md.overlay(new_md, OverlayMode.overwrite)` should instead use `md = new_md.copy()`

mizaki replied:
Because of the commandline clear `-m`, copy can't be used but I'll get into that later. It should be clearer now in the code also.

lordwelch replied:
All you've done is implement a type conversion as an overlay mode. Instead do the type conversion in the overlay method or do a manual check when the metadata from the `-m` option is applied

lordwelch commented on tests/genericmetadata_test.py:
```diff
@@ -22,6 +23,78 @@ def test_metadata_overlay(md: comicapi.genericmetadata.GenericMetadata, replaced
+    assert md == overwrite
+
+
+def test_metadata_overlay_overwrite_no_clear():
+    md = comicapi.genericmetadata.GenericMetadata(
+        series="test", issue="1", title="test", genres={"test", "test2"}, story_arcs=["arc"]
+    )
+    overwrite = comicapi.genericmetadata.GenericMetadata(
+        series="test2", issue="2", title="test2", genres=set(), story_arcs=[]
+    )
+    expected = comicapi.genericmetadata.GenericMetadata(
+        series="test2", issue="2", title="test2", genres={"test", "test2"}, story_arcs=["arc"]
+    )
+    md.overlay(overwrite, OverlayMode.overwrite_no_clear)
+
+    assert md == expected
```
This test needs to show how it is different from overlay. I don't think it ever makes sense to use `overwrite_no_clear` as it is written

mizaki replied:
Now it should show how it's different with comes back to the commandline clear again

lordwelch replied:
This doesn't show all of the types in play.
```python
def test_metadata_overlay_overwrite_no_clear():
    md = comicapi.genericmetadata.GenericMetadata(
        series="test", title="test", issue="1", volume_count=1, page_count=3, day=3, genres={"test", "test2"}, story_arcs=["arc"], characters={"bob", "fred"}
    )
    overwrite = comicapi.genericmetadata.GenericMetadata(
        series="test2", title="", issue_count=2, page_count=2, day=0, genres={'fake'}, characters="", scan_info="nothing",
    )
    expected = comicapi.genericmetadata.GenericMetadata(
        series="test2",
        title="test",
        issue="1",
        issue_count=2,
        volume_count=1,
        day=0,
        page_count=2,
        genres={"fake"},
        story_arcs=["arc"],
        characters={"bob", "fred"},
        scan_info="nothing",
    )
    md.overlay(overwrite, OverlayMode.overwrite_no_clear)

    assert md == expected
```

With this in place we test far more.

- If a `str` is set on both the second value is used: `series`
- If a `str` is set on the first and on the second an empty string is set the first is used: `title`
- If a `str` is set on the first and on the second it is __not__ set the first is used: `issue`
- If a `str` is __not__ set on the first and on the second it is set the second is used: `scan_info`
- If a `int` is set on the first and on the second it is __not__ set the first is used: `volume_count`
- If a `int` is set on both the second value is used: `page_count`
- If a `int` is __not__ set on the first and on the second it is set the second is used: `issue_count`
- If a `int` is set on the first and on the second a `0` is set the second is used: `day`
- If a `set` is set on both the second value is used: `genres`
- If a `list` is set on the first and on the second it is __not__ set the first is used: `story_arcs`
- If a `set` is set on the first and on the second an empty string is set the first is used: `characters`

Now that there is a more complete test, where would this be used? Unless I am mistaken all of these cases are taken care of elsewhere in comictagger. In taggerwindow.py `form_to_metadata` utils.xlate is used to ensure empty string values are set to `None` the same is done with numbers except for `critical_rating` which is special cased at the value 0.0 to be set to `None` because the control doesn't allow it being unset. The same is used in comicrack.py `_convert_xml_to_metadata` except for the special case for `critical_rating` (which means that if someone explicitly set the value to 0 ComicTagger cannot handle it properly and will unset it)

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,24 +252,129 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+                setattr(self, cur, new)
+
+    def assign_add_missing(self, cur: str, new: Any) -> None:
+        """Add Missing - Any values that are None or an empty str/list/set, add them from 'new' to 'self'."""
```
```suggestion
        """Add Missing - Any values that are NOT None or an empty str/list/set, add them from 'new' to 'self'."""
```

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,24 +252,129 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+            elif getattr(self, cur) is None:
+                setattr(self, cur, new)
+
+    def assign_combine_new(self, cur: str, new: Any) -> None:
+        """Combine (New) - Combine lists and sets. Replace all None values. Any other values that collide the 'new' value is used (ignoring None and empty strings)."""
+        if new is not None:
+            if isinstance(new, (list, set)) and isinstance(getattr(self, cur), (list, set)):
+                self.assign_dedupe(new, cur, True)
+            elif (
+                isinstance(new, str)
+                and len(new) > 0
+                and (isinstance(getattr(self, cur), str) or getattr(self, cur) is None)
+            ):
+                setattr(self, cur, new)
+            else:
+                # int, float and bool
+                setattr(self, cur, new)
+
+    def assign_combine_cur(self, cur: str, new: Any) -> None:
+        """Combine (Current) - Same as Combine (New) but 'self' value is kept with collisions."""
+        if new is not None:
+            if isinstance(new, (list, set)) and isinstance(getattr(self, cur), (list, set)):
+                self.assign_dedupe(new, cur)
+            elif isinstance(new, str) and len(getattr(self, cur)) == 0:
+                setattr(self, cur, new)
+            elif getattr(self, cur) is None:
+                setattr(self, cur, new)
```
Why are there 2 of these. The only list/set type you account for is a str (because we don't have any others currently) but that means they are all strings! The only thing you are preserving is case. As named this should have the same implementation as overlay but instead of replacing a list it merges them `if new not in getattr(self, 'cur'): getattr(self, 'cur').add(new)` and `if new not in getattr(self, 'cur'): getattr(self, 'cur').append(new)`

mizaki replied:
I've cleaned this up now and use a `dict` to dedupe so it should be clearer.

lordwelch commented on tests/genericmetadata_test.py:
```diff
@@ -22,6 +23,78 @@ def test_metadata_overlay(md: comicapi.genericmetadata.GenericMetadata, replaced
+    assert md == expected
+
+
+def test_metadata_overlay_combine_cur():
+    md = comicapi.genericmetadata.GenericMetadata(
+        series="test", issue="1", title="test", genres={"test", "test2"}, story_arcs=["arc1"]
+    )
+    combine_cur = comicapi.genericmetadata.GenericMetadata(
+        series="test2", title="test2", genres={"test2", "test3", "test4"}, story_arcs=["arc1", "arc2"]
+    )
+    expected = comicapi.genericmetadata.GenericMetadata(
+        series="test", issue="1", title="test", genres={"test", "test2", "test3", "test4"}, story_arcs=["arc1", "arc2"]
+    )
+    md.overlay(combine_cur, OverlayMode.combine_cur)
+
+    assert md == expected
```
No, this is useless they can just reverse the call of `md.overlay(new_md)` to `new_md.overlay(md)`

mizaki replied:
Again, because of the commandline clear?

lordwelch commented:
Also as a nitpick (because I don't remember to do this all the time anyways) you can change:
```python
        if new is not None:
            ...
```
to
```python
    if new is None:
        return None
    ...
```
to remove another indent level.

As it is now it's 10x better than the 8 levels of indent originally.

mizaki commented:
Part of the problem with `overwrite` and `overwrite_no_clear` was me not understanding the issue of a string clearing a list/set.

Currently https://github.com/comictagger/comictagger/blob/09490b8ebf70415211dc856663ea1c1fd258b47b/comictaggerlib/ctsettings/types.py#L86 doesn't do any type conversions which it probably should here? 
https://github.com/comictagger/comictagger/blob/09490b8ebf70415211dc856663ea1c1fd258b47b/comictaggerlib/ctsettings/types.py#L132

Or, the `GenericMetadata.__post_init__` could do checks and conversions?
That way, overlay doesn't need to worry about the wrong type when replacing/adding/merging.

mizaki commented:
I've think I finally see the issue with the `-m` and clearing list/sets. Using the wrong type for the commandline metadata for a list/set does seem the easiest solution. The other option might be: `Filename`, `commandline`, `read style(s)` and `talker` metadata have different `OverlayModes`. Maybe not all need to be a user available settings (but they could).

I've put comments of `Should always be "overlay" mode?` where I thought the option should be static.

Another (bad) idea would be to mark the origin of the data (maybe re-purpose/use TagOrigin) in the `GenericMetadata`.

lordwelch commented:
You use the quirk on the commandline that an empty string value clears a list as a reason for why these options need to exist..... And then you explicitly say that none of the options you created should be used......
```python
        md.overlay(self.config.Runtime_Options__metadata)  # Should always be "overlay" mode?
```

mizaki commented:
> You use the quirk on the commandline that an empty string value clears a list as a reason for why these options need to exist..... And then you explicitly say that none of the options you created should be used......
> 
> ```python
>         md.overlay(self.config.Runtime_Options__metadata)  # Should always be "overlay" mode?
> ```

It depends on the answer to the question of keep using the quirk or not? If it is used then `OverlayMode.overwrite` for example is required for that reason (i.e. having to check if a list is a list). If, say, `TagOrigin` is used to demarcate if the data is from commandline, file name, etc. then `OverlayMode.overwrite` could be the same (and thus removed and do a copy). It could also be that
```python
# finally, use explicit stuff
md.overlay(self.config.Runtime_Options__metadata)
```
Does not use the user available setting and is always `OverlayMode.overlay` with the `list/set == str` in whereas the other modes do not check.

Having the user choose for all four of the origins (file name, read style, commandline and talker) might lead to confusion?

lordwelch commented:
> It depends on the answer to the question of keep using the quirk or not

No. You have several different modes that you have created. When I said to remove them or asked why they exist you explicitly said that they existed because of the quirk and then you explicitly don't use them the only place the quirk happens. Remove the useless modes. It's a stupid quirk that probably shouldn't ever have been added in the first place.

If something needs to change about how metadata is merged from different sources that needs to change where the decision is made, in this case cli.py `create_local_metadata`. Creating an entire complicated system to facilitate a quirk is stupid. It was implemented as a tiny 7 line change in 539aac1307937ed53ad851496fe5b0f288886a66 to allow unsetting the credits #587. If you can turn this into a 50 line change (minus tests) I'll merge it as is but this is over 700 lines changed, most of which is just to satisfy a stupid quirk that doesn't even get executed.

mizaki commented:
I've squished and targeted what is currently already there. The overlay options will only apply to source (talker) information currently. So this option will become more relevant when the multi-read of metadata styles is done.

I have also renamed `Comic Book Lover` to `Metadata Options` and moved those that are to do with metadata into it.

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,96 +250,190 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
-                    setattr(self, cur, new)
+    def norm_fold(self, string: str) -> str:
+        """Normalise and casefold string"""
+        return unicodedata.normalize("NFKD", string).casefold()
```
Lets move this to utils

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,96 +250,190 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+            cur_dict.update(new_dict)
+            return list(cur_dict.values())
+
+        if isinstance(new, set) and isinstance(cur, set):
```
I don't think we need to do the isinstance check on `new'

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -202,96 +250,190 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
             # otherwise, add it!
             else:
-                self.add_credit(c["person"], c["role"], primary)
+                cur_credits.append(Credit(person=c["person"], role=c["role"], primary=primary))
```
All of the credit appends need to do the same checks as self.add_credit


It might be better to merge them and then clear the current credits and just add them at the end
```python
credits = assign_credit_func(cur_credits, new_credits)
self.credits.clear()
for credit in credits:
    self.add_credit(**credit)
```

mizaki replied:
I have kept it a return value so it's not different to the others. I've also removed the empty person to remove role.

As the `add_credit` is done to each `md` I did a dupe check so using `add_credit` isn't used again. I think it makes sense but if you want to still use it, I'll change it.

lordwelch commented on comictaggerlib/ui/settingswindow.ui:
```diff
@@ -401,17 +391,182 @@
+                </size>
+               </property>
+               <property name="toolTip">
+                <string>Overlay - When the new source metadata has non-empty values, overwrite the current metadata
+
+Add Missing - Any values that are empty in the current metadata that are in the new source metadata, add those values
+
+Combine - Combine lists and use the new (non-empty) source metadata for all other values</string>
```
I don't really like the wording here. What do you think of
 
> Overlay - Overlays new metadata on top of the original.
> (Series=batman, Issue=1, Tags=[batman,joker,robin])
> \+ (Series=Batman, Publisher=DC Comics, Tags=[mystery,action])
> = (Series=Batman, Issue=1, Publisher=DC Comics, Tags=[mystery,action])
> 
> Add Missing - Metadata missing from the original is added from the the new metadata.
> (Series=batman, Issue=1, Tags=[batman,joker,robin])
> \+ (Series=Superman, Issue=10, Publisher=DC Comics, Tags=[superman,lois lane,lex luthor])
> = (Series=batman, Issue=1, Publisher=DC Comics, Tags=[batman,joker,robin])
> 
> Combine - Same as Overlay except lists (tags, characters, genres, credits, etc...) are combined.
> (Series=batman, Issue=1, Tags=[batman,joker,robin])
> \+ (Series=Batman, Publisher=DC Comics, Tags=[mystery,action])
> = (Series=Batman, Issue=1, Publisher=DC Comics, Tags=[batman,joker,robin,mystery,action])

But now that's a lot more text....

How about making this a radio box so each item can have it's own description on the screen and we can make the example the tool tip?

mizaki replied:
Examples should make things easier to understand.

I'm thinking of a text box (on the right as there is space) that changes to reflect the current chosen or, a tabbed text box for the options.

This is kind of related but I think it makes sense to have a choice for the read styles and the source data (with commandline always in `overlay` with clearing of lists etc. and having primacy over read styles). So having two combo boxes makes the most sense then rather than radio.

mizaki replied:
This is what I went with. Could possible have it switch tab when a combo box is changed.
![image](https://github.com/comictagger/comictagger/assets/1141189/359fc757-165d-4038-8ea0-6c6b59b4cf99)

lordwelch replied:
Read style and source info need to have similar text as they are the same operation but the metadata comes from a different place

mizaki replied:
What about covering both? Something like:
```
Source Info: Overlays the new 'source info'  metadata on top of the current metadata as taken from the file name parser or read style (unless 'Overwrite metadata was selected)
Read Style: Overlays subsequent metadata after the first read style from the next read style as set by the user
```
The read style text above probably needs some simplification but that general idea?

lordwelch replied:
So we have the origin of the data "read", "source" and then we have "style" indicating there is some sort of style to the reading, then we have "info" which doesn't contribute any real meaning and doesn't follow the other control. They need to have consistent names. Imagine if on the front page we had this
<img width="256" alt="Screenshot 2024-03-17 at 4 19 18 PM" src="https://github.com/comictagger/comictagger/assets/7547075/761308de-b37a-485b-aa1e-70c433639a03">
That just adds confusion, because instead of keeping consistent names for the controls it uses completely different terms. I really care what terminology you use but it needs to be consistent. And if the name can indicate that "source" is from comicvine (or one of the others) then it needs a description as well (either as a tooltip or a label)

mizaki replied:
IIRC you said to use "style" for the read and modify when it's user facing so I did that. "Data Source" IIRC I changed because of the space it takes up in the tagger window. And then in the settings it's still "Metadata Source", so I think that's the root of the problem I made. In the auto-tag window the option is labelled `Overwrite metadata` so maybe removing the word "Source" from the tagger window and simply put "Metadata" is the answer?

The tool tip for both the label and combox for source info is:
```
The operation to perform when overlaying source data (Comic Vine, Metron, GCD, etc.)
```
and for read style(s):
```
The operation to perform when overlaying read styles (ComicRack, ComicBookInfo, etc.)
```
Are they okay as is?

And the general format of using the one tab to explain both "Metadata" and "Read Style(s)" is okay?

lordwelch replied:
You're missing the point. That was just an example. I don't really care what terminology you use, it just needs to be consistent with the control it is next to.

mizaki replied:
I've changed it to "Data Source" to match the tagger window.

I've changed the description to this:
```
Read Style: Overlays all the none empty values of the read style metadata on top of the current metadata (i.e. file name parsing, other read style(s)).
Data Source: Overlays all the none empty values from the data source (e.g. Comic Vine) on top of the current metadata (i.e. file name parsing, read style(s)).
```
I think now that with the wording there is no need for separate descriptions for "Read Style" and "Data Source"?

lordwelch commented on comictaggerlib/ui/settingswindow.ui:
```diff
@@ -450,99 +545,344 @@
+                <item>
+                 <widget class="QPlainTextEdit" name="combineTextEdit">
+                  <property name="plainText">
+                   <string>Same as Overlay except lists (tags, characters, genres, credits, etc.) are combined.
```
This is not accurate with the current code anymore

mizaki replied:
How's this:
```
Combines lists (tags, characters, genres, credits, etc.) replacing any duplicates with the new record. All other non-empty values replace any current.
```

lordwelch replied:
Is there a reason for the code to not match the description?

mizaki replied:
What am I missing?
```
Combines lists (tags, characters, genres, credits, etc.) replacing any duplicates with the new record. All other non-empty values replace any current.
```
I think that is correct? "replacing" maybe isn't the perfect word, "using the new value with any duplicate items"?

lordwelch replied:
The current description says "Same as Overlay except for lists...." but, it isn't, can we just make the code match the description?

mizaki replied:
I'm with you now. It was a 50/50; change the description or change the code and I chose poorly.

If the command line empty string is taken out (presuming that command line metadata will always be used in `overlay` mode) then they look to do the same outside of lists/sets?
```python
    def assign_overlay(self, cur: Any, new: Any) -> Any:
        """Overlay - When the 'new' object has non-None values, overwrite 'cur'(rent) with 'new'."""
        if new is None:
            return cur
        # An empty string can be used to clear a list or set via the commandline
        if isinstance(new, str) and len(new) == 0:
            if isinstance(cur, (list, set)):
                cur.clear()
                return cur
            else:
                return None
        elif isinstance(new, (list, set)) and len(new) == 0:
            return cur
        else:
            return new
```
```python
    def assign_combine(self, cur: Any, new: Any) -> Any:
        """Combine - Combine lists and sets. All non-None values from new replace cur(rent)"""
        if new is None:
            return cur
        if isinstance(new, (list, set)) and isinstance(cur, (list, set)):
            if len(new) > 0 and isinstance(new, list) and isinstance(new[0], Url):
                return list(set(new).union(cur))
            return self.assign_dedupe(new, cur)
        else:
            return new
```
Apologies if I'm missing the obvious.

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -339,19 +477,20 @@ def get_cover_page_index_list(self) -> list[int]:
-
-        if not found:
-            self.credits.append(credit)
+        if person != "":
```
```suggestion
        if person == "":
            return
```

lordwelch commented on comictaggerlib/settingswindow.py:
```diff
@@ -194,6 +194,9 @@ def __init__(
         )
         self.cbFilenameParser.clear()
         self.cbFilenameParser.addItems(utils.Parser)
+        for mem in OverlayMode:
```
```suggestion
        for mode in OverlayMode:
```

lordwelch commented on comictaggerlib/taggerwindow.py:
```diff
@@ -1166,7 +1166,8 @@ def query_online(self, autoselect: bool = False, literal: bool = False) -> None:
                             ),
-                        )
+                        ),
+                        self.config[0].Metadata_Options__source_overlay,  # command line?
```
What's the question here?

mizaki replied:
That was a question I meant to ask: On the command line the styles from the `file` setting can be overridden but not saved. Should overlay do the same?

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -204,96 +252,186 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+        if new is None:
+            return cur
+        if isinstance(new, (list, set)) and isinstance(cur, (list, set)):
+            if len(new) > 0 and isinstance(new, list) and isinstance(new[0], Url):
+                return self.weblinks_dedupe(new, cur)  # type: ignore[arg-type]
+            return self.assign_dedupe(new, cur)
```
This fails when there are no new links and there is a current link. Probably a good test case.

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -204,96 +252,186 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
-                    pass
-                else:
-                    setattr(self, cur, new)
+    def weblinks_dedupe(self, new: list[Url], cur: list[Url]) -> list[Url]:
+        if len(new) == 0:
+            return cur
+        if len(cur) == 0:
+            return new
+
+        new_set: set[str] = {n.url for n in new}
+        cur_set: set[str] = {c.url for c in cur}
+        new_set.update(cur_set)
+
+        return [parse_url(url) for url in new_set]
```
This is mostly unnecessary. The Url object is a tuple and the url property (actually a method) only does basic null checks and concatenation no sanitizing is done therefore this can be deduplicated with a simple set.
```python
list(set(new).union(cur))
```

mizaki committed:
Remove data to test empty string->None for series and issue as an empty string will never make it to genericmetadata now

mizaki committed:
Change settings menu overlay descriptions

mizaki committed:
Change settings menu overlay descriptions

mizaki committed:
GUI overlay settings moved to internal namespace and CLI args added

mizaki committed:
GUI overlay settings moved to internal namespace and CLI args added

mizaki commented:
I've rebased and squished.

I've gone back through and I don't think there are any outstanding (or accidental reversions).

Also changed the descriptions in the setting window which I hope covers it all now.

I've moved the GUI overlay options to the internal namespace and added CLI version that will override in the same way as `--types` does.

An "add missing with combine (lists/sets)" might be another option to add.

The credits might be better as you suggested as a combine current and new and then pipe them all through `add_credit`. The only reasons not to is it's maybe easier to understand as is, switching of if new overwrites or not (but that could be worked around), and then finally the possibility of adding another overlay mode for credits. Although, I have no idea what that might be right now.

And then there is the question of the `tag_origin`. I imagine you'd want any changes to handling the `notes` field to be a separate PR? So, leave it as is? Maybe the read styles could use it as well? Thinking out loud there...

lordwelch commented:
I rebased the merges out so it's only your new commits

mizaki commented:
> I rebased the merges out so it's only your new commits

Sorry, I usually check but obviously didn't.

lordwelch commented on comictaggerlib/cli.py:
```diff
@@ -265,12 +268,12 @@ def create_local_metadata(self, ca: ComicArchive) -> GenericMetadata:
 
-        # finally, use explicit stuff
+        # finally, use explicit stuff (always 'overlay' mode)
         md.overlay(self.config.Runtime_Options__metadata)
```
Set the mode argument explicitly here `mode=comicapi.genericmetadata.OverlayMode.overlay`

mizaki replied:
I tried `mode=comicapi.genericmetadata.OverlayMode.overlay` but `comicapi` isn't imported so I imported `OverlayMode`.

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -354,12 +472,15 @@ def get_cover_page_index_list(self) -> list[int]:
         found = False
         for c in self.credits:
-            if c["person"].casefold() == person.casefold() and c["role"].casefold() == role.casefold():
+            if norm_fold(c["person"]) == norm_fold(person) and norm_fold(c["role"]) == norm_fold(role):
```
The norm_fold on `person` and `role` should happen outside the loop

lordwelch commented on comicapi/genericmetadata.py:
```diff
@@ -214,101 +261,172 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+            return new
+
+        # Create dict for deduplication
+        new_dict: dict[str, Credit] = {f"{norm_fold(n['person'])}_{n['role'].casefold()}": n for n in new}
```
norm_fold the entire string instead of individually

lordwelch reviewed (changes requested):
I still want to look at the settings and what the tests cover but otherwise this should be just about done

mizaki committed:
norm_fold out of loop for add_credit. Explicit overlay mode for CLI metadata.

mizaki committed:
Move some overlay test data to testing/comicdata.py

mizaki commented:
> I still want to look at the settings and what the tests cover but otherwise this should be just about done

I've moved some of the data into `testing/comicdata.py`. I wasn't sure how to do the credits so left it as a possible "tomorrow" job. Some data might be able to be combined to make it less piecemeal.

mizaki committed:
Rename CBL option to Metadata

mizaki commented:
Only other possible change is `tag_origin` to a list?

mizaki committed:
lordwelch rewrite

mizaki committed:
Separate list merge into a separate option (lordwelch)

mizaki committed:
Rebase corrections

mizaki committed:
Add option for merge lists and fix saving overlay options in settings window

mizaki committed:
Remove combine overlay. Alter help messages in settings window and add lists message

mizaki commented:
I've added the "lists" option. I feel like something is missing now but I have no idea what.

Two dropdowns with only two options seems too few!

mizaki commented:
Thinking about "lists", perhaps it should be three options: merge, replace or ignore (leave original data)?

lordwelch commented:
replace is the wrong word. It falls back to whatever the merge mode is

mizaki commented:
"ignore" might be the final option needed for attributes and lists. That way a user can update only attributes or lists. That would bring lists to a select the same as attributes and bring attributes to three options as well.

Shall I add an ignore and convert "lists" options from a `bool`?

lordwelch commented:
I don't see why someone would treat lists like that. Let's wait until someone has a usecase for it
